pkg/compact/downsample/downsample.go

```diff
@@ -0,0 +1,585 @@
+	return ser, nil
+}
+
+func aggrLset(lset labels.Labels, aggr string) labels.Labels {
+	res := make(labels.Labels, len(lset))
+	copy(res, lset)
+
+	for i, l := range res {
+		if l.Name == "__name__" {
+			res[i].Value = fmt.Sprintf("%s$%s", l.Value, aggr)
+		}
+	}
+	return res
+}
+
+func isCounter(lset labels.Labels) bool {
+	metric := lset.Get("__name__")
+	return strings.HasSuffix(metric, "_total") ||
+		strings.HasSuffix(metric, "_bucket") ||
+		strings.HasSuffix(metric, "_sum")
+}
+
+func targetChunkSize(l int) int {
+	c := 1
+	for ; l/c > 250; c++ {
+	}
+	return l / c
+}
+
+type sample struct {
+	t int64
+	v float64
+}
+
+// series is a mutable series that creates XOR chunk with a configurable target amount
+// of samples per chunk.
+type series struct {
+	lset   labels.Labels
+	chunks []chunks.Meta
+
+	l, n       int
+	cur        chunkenc.Chunk
+	app        chunkenc.Appender
+	cmin, cmax int64
+}
+
+// newSeries returns a new series and automatically cuts a new chunk after n samples if
+// n is greater than 0.
+func newSeries(lset labels.Labels, n int) *series {
+	return &series{lset: lset, n: n}
+}
+
+// newSeries returns a new series and automatically cuts a new chunk after n samples if
+// n is greater than 0.
+// It extends the metric name of the label set by the aggregation modifier.
+func newAggrSeries(lset labels.Labels, aggr string, n int) *series {
+	return &series{lset: aggrLset(lset, aggr), n: n}
+}
+
+func (s *series) add(t int64, v float64) {
+	if s.n > 0 && s.l >= s.n {
+		s.cut()
+	}
+	if s.cur == nil {
+		s.cur = chunkenc.NewXORChunk()
+		s.cmin = t
+		s.app, _ = s.cur.Appender()
+	}
+	s.app.Append(t, v)
+	s.cmax = t
+}
+
+func (s *series) cut() {
+	s.chunks = append(s.chunks, chunks.Meta{
+		MinTime: s.cmin,
+		MaxTime: s.cmax,
+		Chunk:   s.cur,
+	})
+	s.cur = nil
+}
+
+func (s *series) close() {
+	if s.cur == nil {
+		return
+	}
+	s.chunks = append(s.chunks, chunks.Meta{
+		MinTime: s.cmin,
+		MaxTime: s.cmax,
+		Chunk:   s.cur,
+	})
+}
+
+// downsampleCounter downsamples the data points into the series with the given window.
+// If split is set to false, it will write all result data into a single chunk in the series.
+// It writes a signal sample to the end of each chunk which preservers the last true sample of
+// the input data.
+// If it is called on already downsampled data, the input data must end at such a signaling
+// sample to recursively guarantee the semantics.
+func downsampleCounter(ser *series, data []sample, window int64, split bool) {
+	var (
+		n       = targetChunkSize(len(data))
+		nextT   = int64(-1)
+		lastT   int64
+		lastV   float64
+		counter float64
+	)
+	for i, s := range data {
+		if s.t > nextT {
+			if split && i > 0 && i%n == 0 {
+				// Add signaling sample that indicates what the true last value was
+				// before cutting a new chunk.
+				// For already downsampled series lastV is set to the last value that
+				// series itself added as a signaling sample.
+				ser.add(nextT, lastV)
```

Comment:
I think this is wrong... you have somewhere below the comment:
`// If the timestamp increased, it is not the special last sample.`
But here `nextT` is greater than last sample timestamp.